content/en/blog/_posts/2025-XX-XX-mutable-pv-affinity-alpha/index.md

```diff
@@ -0,0 +1,144 @@
+This field was previously immutable,
+and it is now mutable in Kubernetes v1.35 (alpha). This change opens a door to more flexible online volume management.
+
+## Why make node affinity mutable?
+
+So why bother to make it mutable after 8 years?
+While stateless workloads like Deployments can be changed freely
+and the changes will be rolled out automatically by re-creating every Pod,
+PersistentVolumes (PVs) are stateful and cannot be re-created easily without losing data.
+
+However, Storage providers evolve and storage requirements change.
+Most notably, multiple providers are offering regional disks now.
+Some of them even support live migration from zonal to regional disks, without disrupting the workloads.
+This change can be expressed through the recently GA'ed
+[VolumeAttributesClass](https://kubernetes.io/docs/concepts/storage/volume-attributes-classes/) API.
+However, even if the volume is migrated to regional storage,
+Kubernetes still prevents scheduling Pods to other zones because of the node affinity recorded in the PV object.
+In this case, you may want to change the PV node affinity from:
+```yaml
+spec:
+  nodeAffinity:
+    required:
+      nodeSelectorTerms:
+      - matchExpressions:
+        - key: topology.kubernetes.io/zone
+          operator: In
+          values:
+          - us-east1-b
+```
+to:
+```yaml
+spec:
+  nodeAffinity:
+    required:
+      nodeSelectorTerms:
+      - matchExpressions:
+        - key: topology.kubernetes.io/region
+          operator: In
+          values:
+          - us-east1
+```
+
+As another example, providers sometimes offer new generations of disks.
+New disks cannot always be attached to older nodes in the cluster.
+This accessibility can also be expressed through PV node affinity and ensures the Pods can be scheduled to the right nodes.
+But when the disk is upgraded, new Pods using this disk can still be scheduled to older nodes.
+To prevent this, you may want to change the PV node affinity from:
+```yaml
+spec:
+  nodeAffinity:
+    required:
+      nodeSelectorTerms:
+      - matchExpressions:
+        - key: provider.com/disktype.gen1
+          operator: In
+          values:
+          - available
+```
+to:
+```yaml
+spec:
+  nodeAffinity:
+    required:
+      nodeSelectorTerms:
+      - matchExpressions:
+        - key: provider.com/disktype.gen2
+          operator: In
+          values:
+          - available
+```
+
+So, it is mutable now, a first step towards a more flexible online volume management.
+While it is a simple change that removes one validation from the API server,
+we still have a long way to go to integrate well with the Kubernetes ecosystem.
+
+## Try it out
+
+This feature is for you if you are a Kubernetes cluster administrator,
+and your storage provider allows online update that you want to utilize,
+but those updates can affect the accessibility of the volume.
+
```

Comment:
We should add a clear warning here that most existing drivers don't support this feature, and changing the node affinity for a volume in a way that doesn't match what the driver actually supports will create problems.

I like the explanation that this is targeted at administrators who understand that their SPs do support the feature, but we should be clear that in the common case there is nothing to experiment with. Most experimentation is likely to be done by developers adding features to their SPs at this stage.

Reply:
At current stage, there is no extension point for CSI drivers yet. So I want to ask the administrators to modify there volumes from SP directly, by-passing CSI, e.g. from the web console. And I think developers of CSI drivers do not have many things to do yet.

Reply:
My expectation is that developers will add proprietary sidecars with proprietary APIs to do the work. This is how some drivers I used to work on prototyped new features. If this turns out to be a successful approach the new CSI RPCs could be blessed as official, and the proprietary sidecar could be donated to the community.